kivy/tests/test_animations.py

```diff
@@ -35,11 +35,18 @@ def assert_(self, n_start, n_progress_greater_than_zero, n_complete):
     return EventCounter
 
 
+@pytest.fixture(scope='module')
+def approx():
+    from functools import partial
+    return partial(pytest.approx, abs=1)
```

Comment:
Hmm, do we need to make this a fixture rather than simply a global function? But also, why 1? Doesn't the tests use `sleep` to make sure we waited long enough, or there a problem with it?

The canonical way would be to mock the clock, but I'm confused why there would be any timing issues, given that we wait for a given time.

Reply:
> do we need to make this a fixture rather than simply a global function?

The reason I made it a fixture is to avoid importing a module on top-level. (`from functools import partial`). I was told to not do it, as it might have side effects to tests.

> But also, why 1?

There is no reason. I'd appreciate if you tell me what the appropriate value is.

> why there would be any timing issues, given that we wait for a given time.

I have no idea. It's really strange. For example the error in [#7203](https://github.com/kivy/kivy/runs/1366986755):

```
        # during the second half of the second round of the animation
        ec.assert_(1, True, 0)
>       ec1.assert_(2, True, 2)

kivy/tests/test_animations.py:306: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <kivy.tests.test_animations.ec_cls.<locals>.EventCounter object at 0x109d527f0>
n_start = 2, n_progress_greater_than_zero = True, n_complete = 2

    def assert_(self, n_start, n_progress_greater_than_zero, n_complete):
        assert self.n_start == n_start
        if n_progress_greater_than_zero:
            assert self.n_progress > 0
        else:
            assert self.n_progress == 0
>       assert self.n_complete == n_complete
E       assert 1 == 2
E        +  where 1 = <kivy.tests.test_animations.ec_cls.<locals>.EventCounter object at 0x109d527f0>.n_complete
```

this error means the progress of an animation was *behind* than the expectation. If it were *ahead*, I wouldn't be surprised. But it were "behind" for some reason.

Reply:
@matham Don't you think `pytest.approx()` by default is too tight?

```python
assert 100 == pytest.approx(100.5)  # => False
```

This is why I give it `abs=1`

Reply:
>The reason I made it a fixture is to avoid importing a module on top-level. (from functools import partial). I was told to not do it, as it might have side effects to tests.

I did mention this before, but only for kivy modules that have side effects like import kivy logger or widget etc. But it's not a problem for normal python imports which should happen at the top.

>Don't you think pytest.approx() by default is too tight?

Hmm, perhaps, but that feels like a band-aid without fixing the underlying problem. We should first know why 1 sec is too tight, because presumably whoever wrote it didn't think it should be an issue. And once we know why it may be too tight, then we can perhaps increase the interval.

Reply:
ok I'll remove the fixture for now. If its tightness becomes a problem, I might look into it.